File: scripts/test-omnihuman.py
"""
Test script for ElevenLabs Omnihuman 1.5 lip-sync video generation
Uses the official Python SDK

Usage:
    pip install elevenlabs
    python scripts/test-omnihuman.py
"""

import os
import sys
from pathlib import Path

# Load environment variables from .env.local and .env
from dotenv import load_dotenv
load_dotenv('.env.local')
load_dotenv('.env')

ELEVENLABS_API_KEY = os.getenv('ELEVENLABS_API_KEY')
KELLY_VOICE_ID = 'wAdymQH5YucAkXwmrdL0'

# Test text for Kelly to speak
TEST_TEXT = "Hello! I'm Kelly, and I'm so excited to learn with you today! Let's discover something amazing together."

# Output directory
OUTPUT_DIR = Path('test-output')


def test_with_sdk():
    """Test using the official ElevenLabs Python SDK"""
    try:
        from elevenlabs.client import ElevenLabs
        from elevenlabs import play, save
    except ImportError:
        print("❌ ElevenLabs SDK not installed. Run: pip install elevenlabs")
        return False
    
    print("\n🔌 Using ElevenLabs Python SDK...")
    
    # Initialize client
    client = ElevenLabs(api_key=ELEVENLABS_API_KEY)
    
    # Check what attributes are available
    print("\n📋 Available SDK methods:")
    for attr in dir(client):
        if not attr.startswith('_'):
            print(f"   - {attr}")
    
    # Check if image_to_video is available
    if hasattr(client, 'image_to_video'):
        print("\n✅ client.image_to_video is available!")
        print("   Methods:", dir(client.image_to_video))
    else:
        print("\n❌ client.image_to_video NOT available in SDK")
        return False
    
    # Load image
    image_path = Path('public/kelly/poses/kelly_welcome.png')
    if not image_path.exists():
        print(f"❌ Image not found: {image_path}")
        return False
    
    with open(image_path, 'rb') as f:
        image_data = f.read()
    print(f"✅ Image loaded: {len(image_data)} bytes")
    
    # First generate TTS audio
    print("\n📢 Generating TTS audio...")
    try:
        audio = client.text_to_speech.convert(
            voice_id=KELLY_VOICE_ID,
            text=TEST_TEXT,
            model_id="eleven_multilingual_v2"
        )
        
        # Save audio
        audio_path = OUTPUT_DIR / 'kelly-audio.mp3'
        with open(audio_path, 'wb') as f:
            for chunk in audio:
                f.write(chunk)
        
        print(f"✅ Audio saved: {audio_path}")
        
        # Read back for video generation
        with open(audio_path, 'rb') as f:
            audio_data = f.read()
    except Exception as e:
        print(f"❌ TTS error: {e}")
        return False
    
    # Try to generate video
    print("\n🎬 Attempting video generation...")
    
    # Try different method names
    methods_to_try = ['create', 'generate', 'convert']
    
    for method_name in methods_to_try:
        if hasattr(client.image_to_video, method_name):
            print(f"\n🔄 Trying client.image_to_video.{method_name}()...")
            method = getattr(client.image_to_video, method_name)
            
            try:
                result = method(
                    image=image_data,
                    audio=audio_data
                )
                
                print(f"✅ Success! Result type: {type(result)}")
                
                # Save the video
                video_path = OUTPUT_DIR / 'kelly-talking.mp4'
                
                if hasattr(result, 'content'):
                    with open(video_path, 'wb') as f:
                        f.write(result.content)
                elif hasattr(result, 'read'):
                    with open(video_path, 'wb') as f:
                        f.write(result.read())
                elif isinstance(result, bytes):
                    with open(video_path, 'wb') as f:
                        f.write(result)
                else:
                    # Might be a generator
                    with open(video_path, 'wb') as f:
                        for chunk in result:
                            f.write(chunk)
                
                print(f"✅ Video saved: {video_path}")
                return True
                
            except Exception as e:
                print(f"   ❌ Error: {e}")
        else:
            print(f"   ⚠️ Method {method_name} not available")
    
    return False


def test_direct_api():
    """Test using direct API calls with requests"""
    import requests
    
    print("\n🌐 Testing direct API calls...")
    
    # Test user info first
    headers = {'xi-api-key': ELEVENLABS_API_KEY}
    
    resp = requests.get('https://api.elevenlabs.io/v1/user', headers=headers)
    if resp.ok:
        user = resp.json()
        print(f"✅ User: {user.get('subscription', {}).get('tier')}")
        print(f"   Characters: {user.get('subscription', {}).get('character_count')}/{user.get('subscription', {}).get('character_limit')}")
    
    # Check models
    resp = requests.get('https://api.elevenlabs.io/v1/models', headers=headers)
    if resp.ok:
        models = resp.json()
        print(f"\n📋 Available models ({len(models)}):")
        for model in models:
            name = model.get('name', model.get('model_id', 'Unknown'))
            print(f"   - {name}")
            if 'video' in name.lower() or 'omnihuman' in name.lower():
                print(f"     ⭐ Potential video model: {model}")
    
    # Try various video endpoints
    endpoints = [
        'https://api.elevenlabs.io/v1/image-to-video',
        'https://api.elevenlabs.io/v1/video/generate',
        'https://api.elevenlabs.io/v1/videos',
        'https://api.elevenlabs.io/v1/studio/video',
        'https://api.elevenlabs.io/v1/talking-head',
        'https://api.elevenlabs.io/v1/avatar',
    ]
    
    print("\n🔍 Probing video endpoints...")
    for endpoint in endpoints:
        # Try OPTIONS to see if endpoint exists
        resp = requests.options(endpoint, headers=headers)
        get_resp = requests.get(endpoint, headers=headers)
        
        status = get_resp.status_code
        status_str = "✅ Available" if status == 200 else "❓ POST-only" if status == 405 else "❌ Not found" if status == 404 else f"? {status}"
        print(f"   {endpoint.replace('https://api.elevenlabs.io/v1/', '')}: {status_str}")


def list_sdk_structure():
    """List the full structure of the ElevenLabs SDK"""
    try:
        import elevenlabs
        print("\n📦 ElevenLabs SDK structure:")
        print(f"   Version: {elevenlabs.__version__ if hasattr(elevenlabs, '__version__') else 'Unknown'}")
        
        for attr in sorted(dir(elevenlabs)):
            if not attr.startswith('_'):
                obj = getattr(elevenlabs, attr)
                print(f"   {attr}: {type(obj).__name__}")
                
    except ImportError:
        print("❌ SDK not installed")


def main():
    print("═" * 60)
    print("   🎬 ELEVENLABS OMNIHUMAN TEST (Python)")  
    print("═" * 60)
    
    if not ELEVENLABS_API_KEY:
        print("❌ ELEVENLABS_API_KEY not set")
        sys.exit(1)
    
    print(f"✅ API key: {ELEVENLABS_API_KEY[:10]}...")
    
    # Create output directory
    OUTPUT_DIR.mkdir(exist_ok=True)
    print(f"📁 Output: {OUTPUT_DIR.absolute()}")
    
    # List SDK structure
    list_sdk_structure()
    
    # Test direct API
    test_direct_api()
    
    # Test with SDK
    test_with_sdk()
    
    print("\n" + "═" * 60)
    print("   Test complete")
    print("═" * 60)


if __name__ == '__main__':
    main()

File: scripts/test-omnihuman.ts
/**
 * Test script for ElevenLabs Omnihuman 1.5 lip-sync video generation
 * 
 * This script:
 * 1. Reads a Kelly static image
 * 2. Generates TTS audio using ElevenLabs
 * 3. Calls the Omnihuman API to create a lip-synced video
 * 4. Saves the video locally
 * 
 * Usage:
 *   npx ts-node scripts/test-omnihuman.ts
 * 
 * Prerequisites:
 *   - ELEVENLABS_API_KEY environment variable set
 */

import * as dotenv from 'dotenv';
dotenv.config({ path: '.env.local' });
dotenv.config();

import * as fs from 'fs';
import * as path from 'path';

const ELEVENLABS_API_KEY = process.env.ELEVENLABS_API_KEY;
const KELLY_VOICE_ID = 'wAdymQH5YucAkXwmrdL0';

// Test text for Kelly to speak
const TEST_TEXT = "Hello! I'm Kelly, and I'm so excited to learn with you today! Let's discover something amazing together.";

// Output directory
const OUTPUT_DIR = path.join(process.cwd(), 'test-output');

/**
 * Generate TTS audio using ElevenLabs
 */
async function generateTTSAudio(text: string): Promise<Buffer> {
  console.log('📢 Generating TTS audio...');
  console.log(`   Text: "${text.substring(0, 50)}..."`);
  
  const response = await fetch(
    `https://api.elevenlabs.io/v1/text-to-speech/${KELLY_VOICE_ID}`,
    {
      method: 'POST',
      headers: {
        'Accept': 'audio/mpeg',
        'Content-Type': 'application/json',
        'xi-api-key': ELEVENLABS_API_KEY!,
      },
      body: JSON.stringify({
        text,
        model_id: 'eleven_multilingual_v2',
        voice_settings: {
          stability: 0.5,
          similarity_boost: 0.75,
          style: 0.0,
          use_speaker_boost: true,
        },
      }),
    }
  );

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`TTS API error ${response.status}: ${errorText}`);
  }

  const audioBuffer = await response.arrayBuffer();
  console.log(`   ✅ Audio generated: ${audioBuffer.byteLength} bytes`);
  
  return Buffer.from(audioBuffer);
}

/**
 * Test different potential ElevenLabs video API endpoints
 */
async function testVideoEndpoints(imageBuffer: Buffer, audioBuffer: Buffer): Promise<{endpoint: string, result: any}[]> {
  const results: {endpoint: string, result: any}[] = [];
  
  // List of potential endpoints to test
  const potentialEndpoints = [
    'https://api.elevenlabs.io/v1/image-to-video',
    'https://api.elevenlabs.io/v1/image-to-video/omnihuman',
    'https://api.elevenlabs.io/v1/omnihuman/generate',
    'https://api.elevenlabs.io/v1/text-to-video',
    'https://api.elevenlabs.io/v1/avatar/generate',
  ];

  for (const endpoint of potentialEndpoints) {
    console.log(`\n🔍 Testing endpoint: ${endpoint}`);
    
    try {
      // Try with FormData (multipart)
      const formData = new FormData();
      formData.append('source_image', new Blob([imageBuffer], { type: 'image/png' }), 'kelly.png');
      formData.append('audio', new Blob([audioBuffer], { type: 'audio/mpeg' }), 'speech.mp3');
      
      const response = await fetch(endpoint, {
        method: 'POST',
        headers: {
          'xi-api-key': ELEVENLABS_API_KEY!,
        },
        body: formData,
      });

      const contentType = response.headers.get('content-type');
      let result;
      
      if (contentType?.includes('application/json')) {
        result = await response.json();
      } else if (contentType?.includes('video')) {
        result = { type: 'video', size: (await response.arrayBuffer()).byteLength };
      } else {
        result = await response.text();
      }

      console.log(`   Status: ${response.status}`);
      console.log(`   Response:`, typeof result === 'object' ? JSON.stringify(result, null, 2) : result.substring(0, 200));
      
      results.push({ endpoint, result: { status: response.status, body: result } });
      
      // If we got a 200, this endpoint works!
      if (response.ok) {
        console.log(`   ✅ ENDPOINT WORKS!`);
        return results;
      }
    } catch (error) {
      console.log(`   ❌ Error:`, error instanceof Error ? error.message : error);
      results.push({ endpoint, result: { error: String(error) } });
    }
  }
  
  return results;
}

/**
 * Try the Python SDK style endpoint (based on search results)
 */
async function tryImageToVideo(imageBuffer: Buffer, audioBuffer: Buffer): Promise<Buffer | null> {
  console.log('\n🎬 Attempting image-to-video generation...');
  
  // First, let's check what endpoints are available
  console.log('\n📋 Checking available API endpoints...');
  
  try {
    const docsResponse = await fetch('https://api.elevenlabs.io/v1', {
      headers: { 'xi-api-key': ELEVENLABS_API_KEY! }
    });
    console.log(`   API root response: ${docsResponse.status}`);
  } catch (e) {
    console.log('   Could not reach API root');
  }

  // Try the most likely endpoint based on search results
  const endpoints = [
    {
      url: 'https://api.elevenlabs.io/v1/image-to-video',
      method: 'formdata'
    },
    {
      url: 'https://api.elevenlabs.io/v1/image-to-video/generate',
      method: 'formdata'
    }
  ];

  for (const { url, method } of endpoints) {
    console.log(`\n🔄 Trying: ${url}`);
    
    try {
      const formData = new FormData();
      formData.append('image', new Blob([imageBuffer], { type: 'image/png' }), 'kelly.png');
      formData.append('audio', new Blob([audioBuffer], { type: 'audio/mpeg' }), 'speech.mp3');
      // Optional parameters that might be needed
      formData.append('crop_to_face', 'false');

      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'xi-api-key': ELEVENLABS_API_KEY!,
        },
        body: formData,
      });

      console.log(`   Status: ${response.status}`);
      console.log(`   Headers:`, Object.fromEntries(response.headers.entries()));

      if (response.ok) {
        const contentType = response.headers.get('content-type');
        
        if (contentType?.includes('video') || contentType?.includes('octet-stream')) {
          // Direct video response
          const videoBuffer = Buffer.from(await response.arrayBuffer());
          console.log(`   ✅ Got video directly: ${videoBuffer.length} bytes`);
          return videoBuffer;
        }
        
        // JSON response - might contain video URL or generation ID
        const result = await response.json();
        console.log(`   Response:`, JSON.stringify(result, null, 2));
        
        // If there's a video URL, download it
        if (result.video_url || result.output_url || result.url) {
          const videoUrl = result.video_url || result.output_url || result.url;
          console.log(`   📥 Downloading video from: ${videoUrl}`);
          
          const videoResponse = await fetch(videoUrl);
          if (videoResponse.ok) {
            const videoBuffer = Buffer.from(await videoResponse.arrayBuffer());
            console.log(`   ✅ Downloaded video: ${videoBuffer.length} bytes`);
            return videoBuffer;
          }
        }
        
        // If async, poll for completion
        if (result.generation_id || result.id || result.request_id) {
          const id = result.generation_id || result.id || result.request_id;
          console.log(`   ⏳ Async generation started: ${id}`);
          return await pollForVideo(id);
        }
      } else {
        const errorText = await response.text();
        console.log(`   ❌ Error: ${errorText.substring(0, 500)}`);
      }
    } catch (error) {
      console.log(`   ❌ Request failed:`, error instanceof Error ? error.message : error);
    }
  }

  return null;
}

/**
 * Poll for async video generation completion
 */
async function pollForVideo(generationId: string, maxAttempts = 60): Promise<Buffer | null> {
  console.log(`\n⏳ Polling for video completion...`);
  
  const statusEndpoints = [
    `https://api.elevenlabs.io/v1/image-to-video/${generationId}`,
    `https://api.elevenlabs.io/v1/image-to-video/status/${generationId}`,
    `https://api.elevenlabs.io/v1/generations/${generationId}`,
  ];

  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    console.log(`   Attempt ${attempt + 1}/${maxAttempts}...`);
    
    for (const statusUrl of statusEndpoints) {
      try {
        const response = await fetch(statusUrl, {
          headers: { 'xi-api-key': ELEVENLABS_API_KEY! }
        });

        if (!response.ok) continue;

        const result = await response.json();
        console.log(`   Status: ${result.status}`);

        if (result.status === 'completed' || result.status === 'done' || result.status === 'succeeded') {
          const videoUrl = result.video_url || result.output_url || result.url;
          if (videoUrl) {
            const videoResponse = await fetch(videoUrl);
            if (videoResponse.ok) {
              return Buffer.from(await videoResponse.arrayBuffer());
            }
          }
        }

        if (result.status === 'failed' || result.status === 'error') {
          console.log(`   ❌ Generation failed:`, result.error || result.message);
          return null;
        }

        // Found valid endpoint, break inner loop
        break;
      } catch (e) {
        // Try next endpoint
      }
    }

    // Wait before next poll
    await new Promise(r => setTimeout(r, 5000));
  }

  console.log('   ❌ Polling timeout');
  return null;
}

/**
 * Main test function
 */
async function main() {
  console.log('');
  console.log('═══════════════════════════════════════════════════════════════');
  console.log('   🎬 ELEVENLABS OMNIHUMAN 1.5 TEST');
  console.log('═══════════════════════════════════════════════════════════════');
  console.log('');

  // Check API key
  if (!ELEVENLABS_API_KEY) {
    console.error('❌ ELEVENLABS_API_KEY environment variable not set');
    console.log('   Set it in .env.local or export ELEVENLABS_API_KEY=your_key');
    process.exit(1);
  }
  console.log('✅ API key found:', ELEVENLABS_API_KEY.substring(0, 10) + '...');

  // Create output directory
  if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  }
  console.log(`📁 Output directory: ${OUTPUT_DIR}`);

  // Step 1: Read Kelly image
  console.log('\n📷 Loading Kelly image...');
  const imagePath = path.join(process.cwd(), 'public/kelly/poses/kelly_welcome.png');
  
  if (!fs.existsSync(imagePath)) {
    console.error(`❌ Image not found: ${imagePath}`);
    console.log('   Available poses:');
    const posesDir = path.join(process.cwd(), 'public/kelly/poses');
    if (fs.existsSync(posesDir)) {
      fs.readdirSync(posesDir).forEach(f => console.log(`     - ${f}`));
    }
    process.exit(1);
  }
  
  const imageBuffer = fs.readFileSync(imagePath);
  console.log(`   ✅ Image loaded: ${imageBuffer.length} bytes`);

  // Step 2: Generate TTS audio
  const audioBuffer = await generateTTSAudio(TEST_TEXT);
  
  // Save audio for reference
  const audioPath = path.join(OUTPUT_DIR, 'test-audio.mp3');
  fs.writeFileSync(audioPath, audioBuffer);
  console.log(`   💾 Audio saved: ${audioPath}`);

  // Step 3: Try image-to-video generation
  const videoBuffer = await tryImageToVideo(imageBuffer, audioBuffer);
  
  if (videoBuffer) {
    // Save video
    const videoPath = path.join(OUTPUT_DIR, 'kelly-talking.mp4');
    fs.writeFileSync(videoPath, videoBuffer);
    console.log(`\n✅ SUCCESS! Video saved: ${videoPath}`);
    console.log(`   File size: ${videoBuffer.length} bytes`);
  } else {
    console.log('\n⚠️  Could not generate video with standard endpoints.');
    console.log('   The image-to-video API might:');
    console.log('   1. Require beta access / feature flag');
    console.log('   2. Have a different endpoint structure');
    console.log('   3. Not be available on your plan');
    console.log('');
    console.log('   Let me check what API endpoints are available...');
    
    // List available API endpoints
    await checkAvailableEndpoints();
  }
}

/**
 * Check what ElevenLabs API endpoints are available
 */
async function checkAvailableEndpoints() {
  console.log('\n📋 Checking available ElevenLabs API features...');
  
  const testEndpoints = [
    { url: 'https://api.elevenlabs.io/v1/user', name: 'User Info' },
    { url: 'https://api.elevenlabs.io/v1/voices', name: 'Voices' },
    { url: 'https://api.elevenlabs.io/v1/models', name: 'Models' },
    { url: 'https://api.elevenlabs.io/v1/history', name: 'History' },
    { url: 'https://api.elevenlabs.io/v1/convai/agents', name: 'Conversational AI Agents' },
  ];

  for (const { url, name } of testEndpoints) {
    try {
      const response = await fetch(url, {
        headers: { 'xi-api-key': ELEVENLABS_API_KEY! }
      });
      console.log(`   ${name}: ${response.status === 200 ? '✅' : '❌'} (${response.status})`);
      
      // If it's user info, show subscription details
      if (url.includes('/user') && response.ok) {
        const userData = await response.json();
        console.log(`      Subscription: ${userData.subscription?.tier || 'Unknown'}`);
        console.log(`      Character count: ${userData.subscription?.character_count || 0}`);
        console.log(`      Character limit: ${userData.subscription?.character_limit || 0}`);
      }
    } catch (e) {
      console.log(`   ${name}: ❌ Error`);
    }
  }

  // Try to find video-related endpoints
  console.log('\n🔍 Searching for video-related endpoints...');
  const videoEndpoints = [
    'https://api.elevenlabs.io/v1/image-to-video',
    'https://api.elevenlabs.io/v1/video',
    'https://api.elevenlabs.io/v1/avatar',
    'https://api.elevenlabs.io/v1/talking-head',
    'https://api.elevenlabs.io/v1/lipsync',
    'https://api.elevenlabs.io/v1/omnihuman',
  ];

  for (const url of videoEndpoints) {
    try {
      const response = await fetch(url, {
        method: 'GET',
        headers: { 'xi-api-key': ELEVENLABS_API_KEY! }
      });
      const status = response.status;
      const statusText = status === 404 ? 'Not found' : 
                         status === 405 ? 'Method not allowed (POST only?)' :
                         status === 401 ? 'Unauthorized' :
                         status === 403 ? 'Forbidden (need higher tier?)' :
                         status === 200 ? 'Available!' : `Status ${status}`;
      console.log(`   ${url.replace('https://api.elevenlabs.io/v1/', '')}: ${statusText}`);
    } catch (e) {
      console.log(`   ${url}: Error`);
    }
  }
}

// Run
main().catch(error => {
  console.error('\n❌ Fatal error:', error);
  process.exit(1);
});



